README.md
kubernetes-mesos
================

When [Google Kubernetes](https://github.com/GoogleCloudPlatform/kubernetes) meets [Apache Mesos](http://mesos.apache.org/)


[![GoDoc] (https://godoc.org/github.com/mesosphere/kubernetes-mesos?status.png)](https://godoc.org/github.com/mesosphere/kubernetes-mesos)

Kubernetes and Mesos are a match made in heaven. Kubernetes enables the Pod (group of co-located containers) abstraction, along with Pod labels for service discovery, load-balancing, and replication control. Mesos provides the fine-grained resource allocations for pods across nodes in a cluster, and can make Kubernetes play nicely with other frameworks running on the same cluster resources. With the Kubernetes framework for Mesos, the framework scheduler will register Kubernetes with Mesos, and then Mesos will begin offering Kubernetes sets of available resources from the cluster nodes (slaves/minions). The framework scheduler will match Mesos' resource offers to Kubernetes pods to run, and then send a launchTasks message to the Mesos master, which will forward the request onto the appropriate slaves. The slave will then fetch the kubelet/executor and start running it. The kubelet will pull down the pod, start running it, and send a TASK_RUNNING update back to the Kubernetes framework scheduler. Once the pods are running on the slaves/minions, they can make use of Kubernetes' pod labels to register themselves in etcd for service discovery, load-balancing, and replication control. 

### Roadmap
This is still very much a work-in-progress, but stay tuned for updates as we continue development. If you have ideas or patches, feel free to contribute!

- [x] Launching pods (on local machine)
  1. Implement Kube-scheduler API
  1. Pick a Pod (FCFS), match it to an offer.
  1. Launch it!
  1. Kubelet as Executor+Containerizer
- [x] Pod Labels: for Service Discovery + Load Balancing
- [x] Running multi-node on GCE
- [ ] Replication Control
- [ ] Use resource shapes to schedule pods
- [ ] Even smarter (Marathon-like) scheduling

### Build

**NOTE** Kubernetes for Mesos requires Go 1.2+ and protobuf 2.5.0.

```shell
$ sudo aptitude install golang libprotobuf-dev mercurial

$ cd $GOPATH # If you don't have one, create directory and set GOPATH accordingly.

$ go get github.com/mesos/mesos-go/mesos
$ export GOPATH=$GOPATH:$GOPATH/src/github.com/GoogleCloudPlatform/kubernetes/third_party
$ go get github.com/mesosphere/kubernetes-mesos/kubernetes-mesos # If version.go fails to build, rerun after:
$ ./src/github.com/GoogleCloudPlatform/kubernetes/hack/version-gen.sh
$ go get github.com/mesosphere/kubernetes-mesos/kubernetes-mesos
$ go get github.com/mesosphere/kubernetes-mesos/kubernetes-executor
$ go install github.com/GoogleCloudPlatform/kubernetes/cmd/proxy
```

### Start the framework

Ensure that `$(hostname)` resolves to your host's IP address. Assuming your mesos cluster is started, and the master is running on `127.0.1.1:5050`, then:

```shell
$ ./bin/kubernetes-mesos \
  -machines=$(hostname) \
  -mesos_master=127.0.1.1:5050 \
  -executor_path=$(pwd)/bin/kubernetes-executor \
  -etcd_servers=http://localhost:4001 \
  -proxy_path=$(pwd)/bin/proxy
```

###Launch a Pod

Assuming your framework is running on `localhost:8080`, then:

```shell
$ curl -L http://localhost:8080/api/v1beta1/pods -XPOST -d @examples/pod.json
```

After the pod get launched, you can check it's status via `curl` or your web browser:
```shell
$ curl -L http://localhost:8080/api/v1beta1/pods
```

```json
{
	"kind": "PodList",
	"items": [
		{
			"id": "php",
			"labels": {
				"name": "foo"
			},
			"desiredState": {
				"manifest": {
					"version": "v1beta1",
					"id": "php",
					"volumes": null,
					"containers": [
						{
							"name": "nginx",
							"image": "dockerfile/nginx",
							"ports": [
								{
									"hostPort": 8080,
									"containerPort": 80
								}
							],
							"livenessProbe": {
								"enabled": true,
								"type": "http",
								"httpGet": {
									"path": "/index.html",
									"port": "8080"
								},
								"initialDelaySeconds": 30
							}
						}
					]
				}
			},
			"currentState": {
				"manifest": {
					"version": "",
					"id": "",
					"volumes": null,
					"containers": null
				}
			}
		}
	]
}
```

Or, you can run `docker ps -a` to verify that the example container is running:

```shell
CONTAINER ID        IMAGE                       COMMAND                CREATED             STATUS              PORTS               NAMES
3fba73ff274a        busybox:buildroot-2014.02   sh -c 'rm -f nap &&    57 minutes ago                                              k8s--net--php--9acb0442   
```

### Test

Run test suite with:

```shell
$ go test github.com/mesosphere/kubernetes-mesos/kubernetes-mesos -v
```


debian/changelog
kubernetes-mesos (0.3-1) unstable; urgency=medium

  * Bumping up to kubernetes v0.3 

 -- Don Kjer <don@lindenlab.com>  Fri, 26 Sep 2014 02:48:50 +0000

kubernetes-mesos (0.1-1) unstable; urgency=medium

  * Initial release.

 -- Don Kjer <don@lindenlab.com>  Mon, 22 Sep 2014 23:18:17 +0000


build.sh
#!/bin/bash

cd $GOPATH

go get -d github.com/GoogleCloudPlatform/kubernetes

# Switch to mesosphere checkout. Event channels not yet supported in kubernetes
# proper.
pushd src/github.com/GoogleCloudPlatform/kubernetes
mesosphere_remote=`git remote -v | grep "mesosphere"`
if [ $? -ne 0 ]
then
  git remote add mesosphere git@github.com:mesosphere/kubernetes.git
fi

git fetch
git checkout mesos_provider
git pull mesosphere mesos_provider
popd

if [ -d src/github.com/fsouza/go-dockerclient-copiedstructsa ]
then
  git clone https://github.com/nqn/go-dockerclient-copiedstructs src/github.com/fsouza/go-dockerclient-copiedstructs
fi

pushd src/github.com/mesosphere/kubernetes-mesos
godep restore

popd


pushd src/github.com/GoogleCloudPlatform/kubernetes
git checkout mesos_provider
popd

go install github.com/mesosphere/kubernetes-mesos/kubernetes-mesos
go install github.com/mesosphere/kubernetes-mesos/kubernetes-executor
go install github.com/GoogleCloudPlatform/kubernetes/cmd/proxy


executor/executor.go
package executor

import (
	"encoding/json"
	"time"

	"code.google.com/p/goprotobuf/proto"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet"
	log "github.com/golang/glog"
	"github.com/mesosphere/mesos-go/mesos"
	"gopkg.in/v1/yaml"
)

const (
	defaultChanSize   = 1024
	containerPollTime = 1 * time.Second
	launchGracePeriod = 5 * time.Minute
	maxPods           = 256
)

type kuberTask struct {
	mesosTaskInfo     *mesos.TaskInfo
	containerManifest *api.ContainerManifest
}

// KubernetesExecutor is an mesos executor that runs pods
// in a minion machine.
type KubernetesExecutor struct {
	kl         *kubelet.Kubelet // the kubelet instance.
	updateChan chan kubelet.PodUpdate
	eventChan  chan kubelet.Event
	driver     mesos.ExecutorDriver
	registered bool
	tasks      map[string]*kuberTask
	pods       []kubelet.Pod
}

// New creates a new kubernete executor.
func New(driver mesos.ExecutorDriver, kl *kubelet.Kubelet) *KubernetesExecutor {
	return &KubernetesExecutor{
		kl:         kl,
		updateChan: make(chan kubelet.PodUpdate, defaultChanSize),
		eventChan:  make(chan kubelet.Event, defaultChanSize),
		driver:     driver,
		registered: false,
		tasks:      make(map[string]*kuberTask),
	}
}

// Runkubelet runs the kubelet.
func (k *KubernetesExecutor) RunKubelet() {

	go func() {
		for {
			event := <-k.eventChan
			if event.Status == kubelet.PodFailed {
				log.V(2).Infof("Pod launch failed: %s", event.Message)
				if event.Pod != "" {
					log.Warningf("Pod %s failed to launch", event.Pod)
				}
			}
		}
	}()

	k.kl.Run(k.updateChan, k.eventChan)
}

func (k *KubernetesExecutor) gatherContainerManifests() []api.ContainerManifest {
	var manifests []api.ContainerManifest
	for _, task := range k.tasks {
		manifests = append(manifests, *task.containerManifest)
	}
	return manifests
}

func (k *KubernetesExecutor) checkpoint() {
}

// Registered is called when the executor is successfully registered with the slave.
func (k *KubernetesExecutor) Registered(driver mesos.ExecutorDriver,
	executorInfo *mesos.ExecutorInfo, frameworkInfo *mesos.FrameworkInfo, slaveInfo *mesos.SlaveInfo) {
	log.Infof("Executor %v of framework %v registered with slave %v\n",
		executorInfo, frameworkInfo, slaveInfo)
	k.registered = true
}

// Reregistered is called when the executor is successfully re-registered with the slave.
// This can happen when the slave fails over.
func (k *KubernetesExecutor) Reregistered(driver mesos.ExecutorDriver, slaveInfo *mesos.SlaveInfo) {
	log.Infof("Reregistered with slave %v\n", slaveInfo)
	k.registered = true
}

// Disconnected is called when the executor is disconnected with the slave.
func (k *KubernetesExecutor) Disconnected(driver mesos.ExecutorDriver) {
	log.Infof("Slave is disconnected\n")
	k.registered = false
}

// LaunchTask is called when the executor receives a request to launch a task.
func (k *KubernetesExecutor) LaunchTask(driver mesos.ExecutorDriver, taskInfo *mesos.TaskInfo) {
	log.Infof("Launch task %v\n", taskInfo)

	if !k.registered {
		log.Warningf("Ignore launch task because the executor is disconnected\n")
		k.sendStatusUpdate(taskInfo.GetTaskId(),
			mesos.TaskState_TASK_FAILED, "Executor not registered yet")
		return
	}

	taskId := taskInfo.GetTaskId().GetValue()
	if _, found := k.tasks[taskId]; found {
		log.Warningf("Task already launched\n")
		// Not to send back TASK_RUNNING here, because
		// may be duplicated messages or duplicated task id.
		return
	}

	// Get the container manifest from the taskInfo.
	var manifest api.ContainerManifest
	if err := yaml.Unmarshal(taskInfo.GetData(), &manifest); err != nil {
		log.Warningf("Failed to extract yaml data from the taskInfo.data %v\n", err)
		k.sendStatusUpdate(taskInfo.GetTaskId(),
			mesos.TaskState_TASK_FAILED, "Failed to extract yaml data")
		return
	}

	// TODO(nnielsen): Verify this assumption. Manifest ID's has been marked
	// to be deprecated.
	podID := manifest.ID

	// Add the task.
	k.tasks[taskId] = &kuberTask{
		mesosTaskInfo:     taskInfo,
		containerManifest: &manifest,
	}

	k.pods = append(k.pods, kubelet.Pod{
		Name:      podID,
		Namespace: "etcd",
		Manifest:  manifest,
	})

	getPidInfo := func(name string) (api.PodInfo, error) {
		podFullName := kubelet.GetPodFullName(&kubelet.Pod{Name: name, Namespace: "etcd"})

		info, err := k.kl.GetPodInfo(podFullName)
		if err == kubelet.ErrNoContainersInPod {
			return nil, err
		}

		return info, nil
	}

	// TODO(nnielsen): Fail if container is already running.

        // Checkpoint pods.


	// Send the pod updates to the channel.
	// TODO(yifan): Replace SET with ADD when it's implemented.
	// TODO(nnielsen): Incoming launch requests end up destroying already
	// running pods.
	update := kubelet.PodUpdate{
		Pods: k.pods,
		Op:   kubelet.SET,
	}
	k.updateChan <- update

	// Delay reporting 'task running' until container is up.
	go func() {
		expires := time.Now().Add(launchGracePeriod)
		for {
			now := time.Now()
			if now.After(expires) {
				log.Warningf("Launch expired grace period of '%v'", launchGracePeriod)
				break
			}

			// We need to poll the kublet for the pod state, as
			// there is no existing event / push model for this.
			time.Sleep(containerPollTime)

			info, err := getPidInfo(podID)
			if err != nil {
				continue
			}

			log.V(2).Infof("Found pod info: '%v'", info)
			data, err := json.Marshal(info)

			statusUpdate := &mesos.TaskStatus{
				TaskId:  taskInfo.GetTaskId(),
				State:   mesos.NewTaskState(mesos.TaskState_TASK_RUNNING),
				Message: proto.String("Pod '" + podID + "' is running"),
				Data:    data,
			}

			if err := k.driver.SendStatusUpdate(statusUpdate); err != nil {
				log.Warningf("Failed to send status update%v, %v", err)
			}

			// TODO(nnielsen): Monitor health of container and report if lost.
			go func() {
				for {
					time.Sleep(containerPollTime)
					_, err := getPidInfo(podID)
					if err != nil {
						k.sendStatusUpdate(taskInfo.GetTaskId(), mesos.TaskState_TASK_LOST, "Task lost: container disappeared")
						return
					}
				}
			}()

			return
		}
		k.sendStatusUpdate(taskInfo.GetTaskId(), mesos.TaskState_TASK_LOST, "Task lost: launch failed")
	}()
}

// KillTask is called when the executor receives a request to kill a task.
func (k *KubernetesExecutor) KillTask(driver mesos.ExecutorDriver, taskId *mesos.TaskID) {
	log.Infof("Kill task %v\n", taskId)

	if !k.registered {
		log.Warningf("Ignore kill task because the executor is disconnected\n")
		return
	}

	tid := taskId.GetValue()
	if _, ok := k.tasks[tid]; !ok {
		log.Infof("Failed to kill task, unknown task %v\n", tid)
		return
	}
	delete(k.tasks, tid)

	// Send the pod updates to the channel.
	// TODO(yifan): Replace SET with REMOVE when it's implemented.
	update := kubelet.PodUpdate{
		Pods: []kubelet.Pod{},
		Op:   kubelet.SET,
	}
	k.updateChan <- update
	// TODO(yifan): Check the result of the kill event.

	k.sendStatusUpdate(taskId, mesos.TaskState_TASK_KILLED, "Task killed")
}

// FrameworkMessage is called when the framework sends some message to the executor
func (k *KubernetesExecutor) FrameworkMessage(driver mesos.ExecutorDriver, message string) {
	log.Infof("Receives message from framework %v\n", message)
	// TODO(yifan): Check for update message.
}

// Shutdown is called when the executor receives a shutdown request.
func (k *KubernetesExecutor) Shutdown(driver mesos.ExecutorDriver) {
	log.Infof("Shutdown the executor\n")

	for tid, task := range k.tasks {
		delete(k.tasks, tid)

		// Send the pod updates to the channel.
		// TODO(yifan): Replace SET with REMOVE when it's implemented.
		update := kubelet.PodUpdate{
			Pods: []kubelet.Pod{},
			Op:   kubelet.SET,
		}
		k.updateChan <- update
		// TODO(yifan): Check the result of the kill event.

		k.sendStatusUpdate(task.mesosTaskInfo.GetTaskId(),
			mesos.TaskState_TASK_KILLED, "Executor shutdown")
	}
}

// Error is called when some error happens.
func (k *KubernetesExecutor) Error(driver mesos.ExecutorDriver, message string) {
	log.Errorf("Executor error: %v\n", message)
}

func (k *KubernetesExecutor) sendStatusUpdate(taskId *mesos.TaskID, state mesos.TaskState, message string) {
	statusUpdate := &mesos.TaskStatus{
		TaskId:  taskId,
		State:   &state,
		Message: proto.String(message),
	}
	// TODO(yifan): Maybe try to resend again in the future.
	if err := k.driver.SendStatusUpdate(statusUpdate); err != nil {
		log.Warningf("Failed to send status update%v, %v", err)
	}
}


kubernetes-mesos/main.go
/*
Copyright 2014 Google Inc. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// apiserver is the main api server and master for the cluster.
// it is responsible for serving the cluster management API.
package main

import (
	"flag"
	"fmt"
	"net"
	"net/http"
	"strconv"
	"strings"
	"time"

	"code.google.com/p/goprotobuf/proto"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/apiserver"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/master"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/binding"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/controller"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/etcd"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/minion"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/pod"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/latest"
	endpoint "github.com/GoogleCloudPlatform/kubernetes/pkg/service"

	kscheduler "github.com/GoogleCloudPlatform/kubernetes/pkg/scheduler"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
	plugin "github.com/GoogleCloudPlatform/kubernetes/plugin/pkg/scheduler"
	goetcd "github.com/coreos/go-etcd/etcd"
	log "github.com/golang/glog"
	"github.com/mesos/mesos-go/mesos"
	kmscheduler "github.com/mesosphere/kubernetes-mesos/scheduler"
)

var (
	port                        = flag.Uint("port", 8080, "The port to listen on.  Default 8080.")
	address                     = flag.String("address", "127.0.0.1", "The address on the local server to listen to. Default 127.0.0.1")
	apiPrefix                   = flag.String("api_prefix", "/api/v1beta2", "The prefix for API requests on the server. Default '/api/v1beta2'")
	mesosMaster                 = flag.String("mesos_master", "localhost:5050", "Location of leading Mesos master")
	executorPath                = flag.String("executor_path", "", "Location of the kubernetes executor executable")
	proxyPath                   = flag.String("proxy_path", "", "Location of the kubernetes proxy executable")
	minionPort                  = flag.Uint("minion_port", 10250, "The port at which kubelet will be listening on the minions.")
	etcdServerList, machineList util.StringList
)

const (
	artifactPort     = 9000
	cachePeriod      = 10 * time.Second
	syncPeriod       = 30 * time.Second
	httpReadTimeout  = 10 * time.Second
	httpWriteTimeout = 10 * time.Second
)

func init() {
	flag.Var(&etcdServerList, "etcd_servers", "Servers for the etcd (http://ip:port), comma separated")
	flag.Var(&machineList, "machines", "List of machines to schedule onto, comma separated.")
}

type kubernetesMaster struct {
	podRegistry        pod.Registry
	controllerRegistry controller.Registry
	serviceRegistry    service.Registry
	minionRegistry     minion.Registry
	bindingRegistry    binding.Registry
	storage            map[string]apiserver.RESTStorage
	client             *client.Client
	scheduler          *kmscheduler.KubernetesScheduler
}

// Copied from cmd/apiserver.go
func main() {
	flag.Parse()
	util.InitLogs()
	defer util.FlushLogs()

	if len(machineList) == 0 {
		log.Fatal("No machines specified!")
	}

	if len(etcdServerList) <= 0 {
		log.Fatal("No etcd severs specified!")
	}

	serveExecutorArtifact := func(path string) string {
		serveFile := func(pattern string, filename string) {
			http.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
				http.ServeFile(w, r, filename)
			})
		}

		// Create base path (http://foobar:5000/<base>)
		pathSplit := strings.Split(path, "/")
		var base string
		if len(pathSplit) > 0 {
			base = pathSplit[len(pathSplit)-1]
		} else {
			base = path
		}
		serveFile("/"+base, path)

		hostURI := fmt.Sprintf("http://%s:%d/%s", *address, artifactPort, base)
		log.V(2).Infof("Hosting artifact '%s' at '%s'", path, hostURI)

		return hostURI
	}

	executorURI := serveExecutorArtifact(*executorPath)
	proxyURI := serveExecutorArtifact(*proxyPath)

	go http.ListenAndServe(":9000", nil)

	podInfoGetter := &client.HTTPPodInfoGetter{
		Client: http.DefaultClient,
		Port:   *minionPort,
	}

	client, err := client.New("http://"+net.JoinHostPort(*address, strconv.Itoa(int(*port))), "", nil)
	if err != nil {
		log.Fatal(err)
	}

	executorCommand := "./kubernetes-executor -v=2"
	if len(etcdServerList) > 0 {
		etcdServerArguments := strings.Join(etcdServerList, ",")
		executorCommand = "./kubernetes-executor -v=2 -hostname_override=0.0.0.0 -etcd_servers=" + etcdServerArguments
	}

	// Create mesos scheduler driver.
	executor := &mesos.ExecutorInfo{
		ExecutorId: &mesos.ExecutorID{Value: proto.String("KubeletExecutorID")},
		Command: &mesos.CommandInfo{
			Value: proto.String(executorCommand),
			Uris: []*mesos.CommandInfo_URI{
				{Value: &executorURI, Executable: proto.Bool(true)},
				{Value: &proxyURI, Executable: proto.Bool(true)},
			},
		},
		Name:   proto.String("Kubelet Executor"),
		Source: proto.String("kubernetes"),
	}

	etcdClient := goetcd.NewClient(etcdServerList)
	helper := tools.EtcdHelper{
		etcdClient,
		latest.Codec,
		latest.ResourceVersioner,
	}
	mesosPodScheduler := kmscheduler.New(executor, kmscheduler.FCFSScheduleFunc, client, helper)
	driver := &mesos.MesosSchedulerDriver{
		Master: *mesosMaster,
		Framework: mesos.FrameworkInfo{
			Name: proto.String("KubernetesScheduler"),
			User: proto.String("root"),
		},
		Scheduler: mesosPodScheduler,
	}
	mesosPodScheduler.Driver = driver

	driver.Init()
	defer driver.Destroy()
	go driver.Start()

	log.V(2).Info("Serving executor artifacts...")
	// TODO(nnielsen): Using default pod info getter until
	// MesosPodInfoGetter supports network containers.

	// podInfoGetter := MesosPodInfoGetter.New(mesosPodScheduler)

	var cloud cloudprovider.Interface
	m := newKubernetesMaster(mesosPodScheduler, &master.Config{
		Client:        client,
		Cloud:         cloud,
		Minions:       machineList,
		PodInfoGetter: podInfoGetter,
		EtcdHelper:    helper,
	}, helper)
	log.Fatal(m.run(net.JoinHostPort(*address, strconv.Itoa(int(*port))), *apiPrefix, helper.Codec))
}

func newKubernetesMaster(scheduler *kmscheduler.KubernetesScheduler, c *master.Config, etcdHelper tools.EtcdHelper) *kubernetesMaster {
	var m *kubernetesMaster

	minionRegistry := minion.NewRegistry(c.Minions) // TODO(adam): Mimic minionRegistryMaker(c)?

	m = &kubernetesMaster{
		podRegistry:        scheduler,
		controllerRegistry: etcd.NewRegistry(etcdHelper),
		serviceRegistry:    etcd.NewRegistry(etcdHelper),
		minionRegistry:     minionRegistry,
		bindingRegistry:    etcd.NewRegistry(etcdHelper),
		client:             c.Client,
		scheduler:          scheduler,
	}
	m.init(scheduler, c.Cloud, c.PodInfoGetter)

	return m
}

func (m *kubernetesMaster) init(scheduler kscheduler.Scheduler, cloud cloudprovider.Interface, podInfoGetter client.PodInfoGetter) {
	podCache := master.NewPodCache(podInfoGetter, m.podRegistry)
	go util.Forever(func() { podCache.UpdateAllContainers() }, cachePeriod)
	m.storage = map[string]apiserver.RESTStorage{
		"pods": pod.NewREST(&pod.RESTConfig{
			CloudProvider: cloud,
			PodCache:      podCache,
			PodInfoGetter: podInfoGetter,
			Registry:      m.podRegistry,
		}),
		"replicationControllers": controller.NewREST(m.controllerRegistry, m.podRegistry),
		"services":               service.NewREST(m.serviceRegistry, cloud, m.minionRegistry),
		"minions":                minion.NewREST(m.minionRegistry),
		// TODO: should appear only in scheduler API group.
		"bindings": binding.NewREST(m.bindingRegistry),
	}
}

// Run begins serving the Kubernetes API. It never returns.
func (m *kubernetesMaster) run(myAddress, apiPrefix string, codec runtime.Codec) error {
	endpoints := endpoint.NewEndpointController(m.serviceRegistry, m.client)
	go util.Forever(func() { endpoints.SyncServiceEndpoints() }, syncPeriod)
	plugin.New(m.scheduler.NewPluginConfig()).Run()

	s := &http.Server{
		Addr:           myAddress,
		Handler:        apiserver.Handle(m.storage, codec, apiPrefix),
		ReadTimeout:    httpReadTimeout,
		WriteTimeout:   httpWriteTimeout,
		MaxHeaderBytes: 1 << 20,
	}
	return s.ListenAndServe()
}
